File: tests/standardization/test_standardize.c
#include "bayesrco_io.h"
#include "utils.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main() {
    IOConfig ioconfig;
    GenomicData gdata;

    memset(&ioconfig, 0, sizeof(IOConfig));
    memset(&gdata, 0, sizeof(GenomicData));
    
    strcpy(ioconfig.geno_file_path, "test_data.bed");
    strcpy(ioconfig.pheno_file_path, "test_data.fam");
    ioconfig.trait_column_index = 1;
    ioconfig.mcmc = true;

    // Load dimensions
    if (io_get_size(&ioconfig, &gdata) != SUCCESS) {
        fprintf(stderr, "Error in io_get_size\n");
        return 1;
    }
    
    // Load phenotypes (implicitly allocates trains)
    if (io_load_phenotypes(&ioconfig, &gdata) != SUCCESS) {
        fprintf(stderr, "Error in io_load_phenotypes\n");
        return 1;
    }
    
    // Allocate logic for X and frequencies
    int nt = gdata.num_phenotyped_individuals;
    gdata.genotypes = (double*)calloc(gdata.num_loci * nt, sizeof(double));
    gdata.allele_frequencies = (double*)calloc(gdata.num_loci, sizeof(double));
    
    // Load genotypes
    if (io_load_genotypes(&ioconfig, &gdata) != SUCCESS) {
        fprintf(stderr, "Error in io_load_genotypes\n");
        return 1;
    }

    // Run standardization
    standardize_genotypes(&gdata, true);

    printf("Allele Frequencies (freqstore):\n");
    for (int j = 0; j < gdata.num_loci; j++) {
        printf("%25.16E\n", gdata.allele_frequencies[j]);
    }

    printf("Standardized Genotypes (X):\n");
    for (int i = 0; i < nt; i++) {
        for (int j = 0; j < gdata.num_loci; j++) {
            // C genotypes are column-major (nloci x nt), so X[j * nt + i]
            printf("%25.16E\n", gdata.genotypes[j * nt + i]);
        }
    }

    // Cleanup
    io_cleanup(&ioconfig, &gdata, NULL, NULL);

    return 0;
}

File: tests/standardization/test_standardize.f90
program test_standardize
    use mod_defs, only: dp, GENOTYPE_MISSING_THRESHOLD
    use mod_data
    use mod_io
    use mod_standardize
    implicit none

    type(ModelConfig) :: config_local
    type(GenomicData) :: gdata_local
    integer :: i, j

    config_local%inprefix = "test_data"
    config_local%phenfil = "test_data.fam"
    config_local%bimfil = "test_data.bim"
    config_local%genfil = "test_data.bed"
    config_local%freqfil = "test_data.frq"
    config_local%trait_pos = 1
    config_local%mcmc = .true.

    call get_size(config_local, gdata_local)
    call load_phenos_plink(config_local, gdata_local)
    
    ! Allocate X
    gdata_local%nt = count(gdata_local%trains == 0)
    allocate(gdata_local%X(gdata_local%nt, gdata_local%nloci))
    allocate(gdata_local%freqstore(gdata_local%nloci))
    
    call load_snp_binary(config_local, gdata_local)
    
    ! Run xcenter
    call xcenter(config_local, gdata_local)

    print *, "Allele Frequencies (freqstore):"
    do j = 1, gdata_local%nloci
        print "(E25.16)", gdata_local%freqstore(j)
    end do

    print *, "Standardized Genotypes (X):"
    do i = 1, gdata_local%nt
        do j = 1, gdata_local%nloci
            print "(E25.16)", gdata_local%X(i, j)
        end do
    end do

end program test_standardize
